### ISSUE TITLE
Migration crash with AttributeError: type object 'Employee' has no attribute 'objects'

### ISSUE THREAD

**Issue by msgre** [django-mptt/django-mptt#489]
Hello,
we are using `django-mptt` in our project and after some time we found we have broken migrations (we have ~50 of them in `employees` application).

Problem is in data migration. Every time we access `objects` attribute, and try to querying model, we get nasty error message `AttributeError: type object 'Employee' has no attribute 'objects'`. Data migration looks like this:

```
# -*- coding: utf-8 -*-
# Generated by Django 1.9.7 on 2016-08-03 06:48
from __future__ import unicode_literals

from django.db import migrations

def simple_test(apps, schema_editor):
    Employee = apps.get_model("employees", "Employee")
    for employee in Employee.objects.all():
        # ... some data logic here
        employee.save()

class Migration(migrations.Migration):

    dependencies = [
        ('employees', '0001_initial'),
    ]

    operations = [
        migrations.RunPython(simple_test),
    ]
```

Migration fails with this traceback:
```
  Applying employees.0002_auto_20160803_0648...Traceback (most recent call last):
  File "./manage.py", line 18, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/__init__.py", line 353, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.5/site-packages/django/core/management/__init__.py", line 345, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/base.py", line 348, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/base.py", line 399, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/commands/migrate.py", line 200, in handle
    executor.migrate(targets, plan, fake=fake, fake_initial=fake_initial)
  File "/usr/local/lib/python3.5/site-packages/django/db/migrations/executor.py", line 92, in migrate
    self._migrate_all_forwards(plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/usr/local/lib/python3.5/site-packages/django/db/migrations/executor.py", line 121, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/usr/local/lib/python3.5/site-packages/django/db/migrations/executor.py", line 198, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/usr/local/lib/python3.5/site-packages/django/db/migrations/migration.py", line 123, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/usr/local/lib/python3.5/site-packages/django/db/migrations/operations/special.py", line 183, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/path-to-our-project/employees/migrations/0002_auto_20160803_0648.py", line 10, in combine_names
    for employee in Employee.objects.all():
AttributeError: type object 'Employee' has no attribute 'objects'
```

Our `Employee` model is defined like this:

```
from mptt.models import MPTTModel, TreeForeignKey

class Employee(MPTTModel):
    # nunch of attributes...
    supervisor = TreeForeignKey('self', null=True, blank=True, related_name='employees', db_index=True)

    class MPTTMeta:
        parent_attr = 'supervisor'
```

First, I think that we have some problem in our migrations (we have several times some migration conflict, because 4 developers work on this project and sometimes they produce migration files in same time but with different changes inside).
But today I remove all of the migration files, generated just one initial from current state of our app, make cleanup in `django_migrations` DB table and then create dumb data migration file.

But same crash occurs.

After some investinagation I find that if I make following manual modification of initial migration, I am able to access `objects` again and all of migrations run without problem.

Snippet from initial migration:

        migrations.CreateModel(
            name='Employee',
            fields=[
                # list of our attributes, nothing special here...
                ('lft', models.PositiveIntegerField(db_index=True, editable=False)),
                ('rght', models.PositiveIntegerField(db_index=True, editable=False)),
                ('tree_id', models.PositiveIntegerField(db_index=True, editable=False)),
                ('level', models.PositiveIntegerField(db_index=True, editable=False)),
            ],
            options={
                'permissions': # list of permissions
            },
            managers=[
                ('objects', django.db.models.manager.Manager()), # <== THIS IS MY MODIFICATION
                ('_default_manager', django.db.models.manager.Manager()),
            ],
        ),

In other words, I copy `('_default_manager', django.db.models.manager.Manager()),` row and replaced term `_default_manager` with `objects`.

I don't really know if this hack have some side effects, but it works.

Do you know what is going here? Is it `mptt` cause of the problem or not? (I believe it is, because data migration on standard models with `models.Model` base class works without problem). Is there some standard solution?

Thanks!

SW versions: Python 3.5.1, Django 1.9.7, mptt 0.8.4

**Comment by pySilver** [comment]
I have the same issue!

**Comment by craigds** [comment]
Can you try this with just-released 0.8.6? There are some manager changes.

**Comment by ArkmonSpirit** [comment]
I have the same issue too and still the same with version 0.8.6

**Comment by nuarhu** [comment]
This was necessary to get it to work for us:

In one migration file:

    def get_model(apps):
        CategoryTreeNode = apps.get_model('category', 'CategoryTreeNode')
        from mptt import register
        register(CategoryTreeNode)  # this prevents getting constraint violation errors for `lft`
        return CategoryTreeNode

    class Migration(migrations.Migration):
        dependencies = [
            ('category', '0001_initial'),
        ]
    
        operations = [
            # This is required or 'apps.get_model' will fail in the end
            migrations.AlterModelManagers(
                name='categorytreenode',
                managers=[
                ],
            ),
            migrations.RunPython(import_category_data, remove_category_data),
        ]

Using:
Django==1.9.7
django-mptt==0.8.6

**Comment by Rikuoja** [comment]
The same issue here!

django==1.9.11
django-mptt==0.8.6

**Comment by Rikuoja** [comment]
Further, altering the model manager to `django.db.models.manager.Manager()` doesn't help. That results in a sorting problem with the added manager:
``` 
File "/Users/riku/.virtualenvs/linkedevents/lib/python3.4/site-packages/django/db/migrations/state.py", line 537, in <lambda>
    sorted_managers = sorted(self.managers, key=lambda v: v[1].creation_counter)
AttributeError: 'str' object has no attribute 'creation_counter'
```

**Comment by zelds** [comment]
Same issue.

**Comment by erikvw** [comment]
not sure if you have solved this but ... you most likely have a custom model manager. For the "objects" model manager to be accessible in migrations you need to set "use_in_migrations = True" when declaring the custom model manager:

```
class MyCustomManager(models.Manager):

    use_in_migrations = True
    ...
```

Now it should be available.

**Comment by jrobichaud** [comment]
Correct me if I am work bug I believe django do not need the `AlterModelManagers` anymore unless to override managers with `use_in_migrations = True`.

If this is true, the real now question is: 

> Should the MPTT managers be used during migrations?

### If Yes: 

#### Remove all:
```python
migrations.AlterModelManagers(
    name='entity',
    managers=[
        'objects', django.db.models.manager.Manager()),
    ],
),
```
#### and remove these sections of `migrations.CreateModel`:

```python
        managers=[
            ('objects', django.db.models.manager.Manager()), # <== THIS IS MY MODIFICATION
            ('_default_manager', django.db.models.manager.Manager()),
        ],
```

Add `use_in_migrations = True` to the managers like @erikvw said.

### If No:

#### Remove all:
```python
migrations.AlterModelManagers(
    name='entity',
    managers=[
        'objects', django.db.models.manager.Manager()),
    ],
),
```
#### and remove these sections of `migrations.CreateModel`:

```python
        managers=[
            ('objects', django.db.models.manager.Manager()), # <== THIS IS MY MODIFICATION
            ('_default_manager', django.db.models.manager.Manager()),
        ],
```

**Comment by bartektrybala** [comment]
Employee from apps.get_model("employees", "Employee") is fake of your Employee model. Probably you override Employee `objects` property with CustomQuerySet or something like this.  Use this:
`Employee.default_manager.all()`

**Comment by matthiask** [comment]
The workarounds are now documented in the installation docs: use `migrations.AlterModelManagers` to swap out the tree manager before a `RunPython`, or call `mptt.register()` on the historical model inside the migration function. See #833.
